src/goto-programs/goto_convert_side_effect.cpp

```diff
@@ -128,8 +128,12 @@ void goto_convertt::remove_assignment(
   // revert assignment in the expression to its LHS
   if(result_is_used)
   {
-    exprt lhs;
-    lhs.swap(to_binary_expr(expr).op0());
+    // assign_* statements can have an lhs operand with a different type than
+    // that of the overall expression, because of integer promotion (which may
+    // have introduced casts to the lhs). skip over those type casts, but also
+    // make sure the resulting expression has the same type as before.
+    exprt lhs = typecast_exprt::conditional_cast(
+      skip_typecast(to_binary_expr(expr).op0()), expr.type());
     expr.swap(lhs);
```

Comment:
I am a bit worried about the `skip_typecast`, since that may remove any number of casts.

Say think of `(int)(char)i`, where `i` is an integer.

Reply:
Thank you for calling this out! I have now added sanity checks and refrained from using `skip_typecast`.